vxfreeswitch/voice.py

```diff
@@ -131,6 +132,50 @@ def unboundEvent(self, evdata, evname):
         log.msg("Unbound event %r" % (evname,))
 
 
+class FreeSwitchESLClientProtocol(FreeSwitchESLProtocol):
+    def __init__(self, vumi_transport, number):
+        FreeSwitchESLProtocol.__init__(self, vumi_transport)
+        self.uniquecallid = number
+        self.job_queue = {}
+        self.ready = Deferred()
+
+    @inlineCallbacks
+    def connectionMade(self):
+        yield self.eventplain("BACKGROUND_JOB CHANNEL_HANGUP")
+        yield self.vumi_transport.register_client(self)
+        self.ready.callback(self)
+
+    def make_call(self):
+        def _success(ev):
+            response = Deferred()
+            self.job_queue[ev.Job_UUID] = response
+            return response
+        profile = self.vumi_transport.config.sofia_profile
+        call_url = "sofia/%s/%s" % (profile, self.uniquecallid)
+        d = self.bgapi("originate %s" % (call_url))
+        d.addCallback(_success)
+        return d
+
+    def onBackgroundJob(self, ev):
+        d = self.job_queue.pop(ev.Job_UUID, None)
+        if d:
+            response, content = ev.rawresponse.split()
+            if response == "+OK":
+                d.callback(content)
+
+    @inlineCallbacks
+    def onChannelHangup(self, ev):
+        self.vumi_transport.deregister_client(self)
+        yield self.transport.loseConnection()
+
+class DialerFactory(ClientFactory):
+    def __init__(self, vumi_transport, number):
+        self.vumi_transport = vumi_transport
+        self.number = number
+
+    def buildProtocol(self, addr):
+        return FreeSwitchESLClientProtocol(self.vumi_transport, self.number)
```

Comment:
We probably need to do:
```python
p = FreeSwitchESLClientProtocol(self.vumi_transport, self.number)
p.factory = self
return p
```
here because that's what http://twistedmatrix.com/trac/browser/tags/releases/twisted-15.0.0/twisted/internet/protocol.py#L113 does. Alternatively I guess we could just define a method called ``protocol``?